### ISSUE TITLE
`CreatedResult` should accept null location

### ISSUE THREAD

**Issue by aayjaychan** [dotnet/aspnetcore#39454]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

[RFC7231](https://datatracker.ietf.org/doc/html/rfc7231#section-6.3.2) suggests that the `Location` header is optional for `201` Created response.

> The primary resource created by the request is identified by either a Location header field in the response or, if no Location field is received, by the effective request URI.

However, currently the constructors of `CreateResult` throws if `null` is passed to `location`, and `ControllerBase.Created()` throws if `null` is passed to `url`.

### Expected Behavior

A response of  `201` Created without `Location` header is created.

### Steps To Reproduce

```C#
[HttpPost]
public ActionResult Create()
{
    return new CreatedResult((string?)null, null); // throws ArgumentNullException
}
```

```C#
[HttpPost]
public ActionResult Create()
{
    return Created((string?)null, null) // throws ArgumentNullException
}
```

### Exceptions (if any)

```
System.ArgumentNullException: Value cannot be null. (Parameter 'location')
   at Microsoft.AspNetCore.Mvc.CreatedResult..ctor(String location, Object value)
   ...
```

### .NET Version

6.0.101

**Comment by khellang** [comment]
```csharp
[HttpPost]
public ActionResult Create()
{
    return new StatusCodeResult(StatusCodes.Status201Created);
}
```

🤔

**Comment by khellang** [comment]
Since your examples pass `null` as the value instead, you might as well just use `StatusCode` or `StatusCodeResult`.

Now, I have wished for an overload of `Created` that has `value`, but not `uri`. A lot of times when you return back a complete result, it's unnecessary to return the location to the client as that URI would just yield the same result.

**Comment by halter73** [comment]
Strawman API proposal:

```diff
namespace Microsoft.AspNetCore.Mvc
{
    public class CreatedResult : ObjectResult
    {
+        public CreatedResult()
    
-        public CreatedResult(string location, object? value)
+        public CreatedResult(string? location, object? value)

-     public CreatedResult(Uri location, object? value)
+    public CreatedResult(Uri? location, object? value)
    }

    public abstract class ControllerBase
    {
+        public virtual CreatedResult Created()
    
-        public virtual CreatedResult Created(string uri, [ActionResultObjectValue] object? value)
+        public virtual CreatedResult Created(string? uri, [ActionResultObjectValue] object? value)

-         public virtual CreatedResult Created(Uri uri, [ActionResultObjectValue] object? value)
+        public virtual CreatedResult Created(Uri? uri, [ActionResultObjectValue] object? value)
    }
}

namespace Microsoft.AspNetCore.Http
{
    public static class Results
    {
+        public static IResult Created()
    
-        public static IResult Created(string uri, object? value)
+        public static IResult Created(string uri?, object? value)

-        public static IResult Created(Uri uri, object? value)
+        public static IResult Created(string uri?, object? value)
    }
}
```

I'm not super convinced that making the location/uri parameters nullable is a huge win since you'd need either `(string?)null` or `(Uri?)null` to select the right overload. We could also consider adding single parameter overloads that just take the location/uri.

**Comment by khellang** [comment]
It's too bad that `uri` is before `value` already, but I don't think omitting a preceding parameter in one (or some) overloads is _as bad_ as changing order between overloads (with same or more parameters.

> you'd need either `(string?)null` or `(Uri?)null` to select the right overload.

The fact that there's two overloads makes it even more compelling to add a single-parameter overload, since you need those ugly `(string?)null` casts everywhere to avoid ambiguity 😞

**Comment by pranavkm** [comment]
We could consider making only one of the overloads (the `string?` one) and make it nullable. If the compiler supports using nullablility as part of overload resolution, users could get away with writing `Created(null, "My-cool-value")` without having to cast. Outside of that API approved.

**Comment by Marusyk** [comment]
I'm gonna prepare a PR for this, pls assign it to me

**Comment by Marusyk** [comment]
Could you please suggest what is the final decision after API review?

**Comment by halter73** [comment]
> Could you please suggest what is the final decision after API review?

We should have been clearer about it, but the approved API is exactly what's described in https://github.com/dotnet/aspnetcore/issues/39454#issuecomment-1012457232.

> We could consider making only one of the overloads (the string? one) and make it nullable. If the compiler supports using nullablility as part of overload resolution, users could get away with writing Created(null, "My-cool-value") without having to cast. Outside of that API approved.

This is [not supported.](https://sharplab.io/#v2:D4AQTAjAsAULIGYAE4kGEkG9ZNy5IALEgLIAUAlFjnrWgE4CmAhgC6MAmZAdgK4A2/ADRIARCQCeAWgDGAezn8pAN2b9ejURQDcNXAF89SI4hQQAbCmIMW7LiAgAGAPxJe9AJYi5AIwBWjDKsrqrqjBRG2DC0hnDReKYOlkToTGycZACqnm6e3v6BwUihGhHxuFExsPpAA==)

**Comment by singh733** [comment]
Hi @pranavkm @halter73 

Below is the proposed solution:

a) Remove Uri overload of CreatedResult and refactor the  String overload in the CreatedResult.cs.
![image](https://user-images.githubusercontent.com/83835739/172155895-113dfa01-2093-4150-bd1e-b46a6dbb1ae4.png)

b) Remove CreatedResult method from the ControllerBase.
![image](https://user-images.githubusercontent.com/83835739/172156056-9bf54728-dcb9-41e5-aa80-605788b4f32c.png)

c) Remove the following Test cases.

![image](https://user-images.githubusercontent.com/83835739/172156161-3274e58a-d759-4c94-a8ea-bcf390062afe.png)


Please Note:

When I test like below it works.
![image](https://user-images.githubusercontent.com/83835739/172156297-63d34f94-c7d1-4d37-8c77-521be1cc4b01.png)

Please let me know if  this is fine or not. If yes,  then i will send PR

**Comment by halter73** [comment]
The approved API does not remove any constructors or methods. It merely changes the nullability annotations of some parameters.

If you need to remove test cases for them to compile or pass, that's an indication of a [breaking change](https://docs.microsoft.com/en-us/dotnet/core/compatibility/#members) which we will not accept.

> ❌ DISALLOWED: Renaming or removing a public member or parameter
>
> This breaks all code that uses the renamed or removed member, or parameter.

It's important that developers can confidently upgrade between versions of ASP.NET Core without updating a bunch of code. This sometimes means saying no to cleaning up APIs that are good enough because doing so would break already working applications.

> When I test like below it works.
> 
> ```csharp
> [HttpPost]
> public ActionResult Create()
> {
>     return new CreatedResult(null, "Test");
> }
> ```

I recommend using [markdown for code snippets](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code) so we can search the content and more easily quote it.

This can only work because of the breaking change. This won't compile with the approved changes. You'll get the same error demonstrated by the [sharplab.io sample I posted in my last comment on this issue](https://sharplab.io/#v2:D4AQTAjAsAULIGYAE4kGEkG9ZNy5IALEgLIAUAlFjnrWgE4CmAhgC6MAmZAdgK4A2/ADRIARCQCeAWgDGAezn8pAN2b9ejURQDcNXAF89SI4hQQAbCmIMW7LiAgAGAPxJe9AJYi5AIwBWjDKsrqrqjBRG2DC0hnDReKYOlkToTGycZACqnm6e3v6BwUihGhHxuFExsPpAA==). The following will compile though without warnings and work.

```csharp
[HttpPost]
public ActionResult Create()
{
    return new CreatedResult((string?)null, "Test");
}
```

Currently, this would give nullability warnings and throw an `ArgumentNullException` at runtime.

**Comment by singh733** [comment]
@halter73  Thank you for the Feedback !!


> We could consider making only one of the overloads (the `string?` one) and make it nullable. If the compiler supports using nullablility as part of overload resolution, users could get away with writing `Created(null, "My-cool-value")` without having to cast. Outside of that API approved.

If we will do this then we still get this error [sharplab.io.](https://sharplab.io/#v2:D4AQTAjAsAULIGYAE4kGEkG9ZNy5IALEgLIAUAlFjnrWgE4CmAhgC6MAmZAdgK4A2/ADRIARCQCeAWgDGAezn8pAN2b9ejURQDcNXAF89SI4hQQAbCmIMW7LiAgAGAPxJe9AJYi5AIwBWjDKsrqrqjBRG2DC0hnDReKYOlkToTGycZACqnm6e3v6BwUihGhHxuFExsPpAA==)


Proposed Solution:

We can create one more overload like below:

```
public CreatedResult(object? value): base(value)       
{
    StatusCode = DefaultStatusCode;
}
```

Please share your views.

**Comment by halter73** [comment]
We considered the one-parameter overload but decided against it:

> We could also consider adding single parameter overloads that just take the value as @khellang suggested. I think that would be more useful than a single parameter overloads that just take the location/uri, but that would violate the [Member overloading framework design guidelines](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/member-overloading) which is why I omitted it.
> 
> > ❌ AVOID being inconsistent in the ordering of parameters in overloaded members. Parameters with the same name should appear in the same position in all overloads.

**Comment by singh733** [comment]
Please let me know what should we do now ?

**Comment by halter73** [comment]
You can implement the approved API described in https://github.com/dotnet/aspnetcore/issues/39454#issuecomment-1012457232 if you want.

**Comment by singh733** [comment]
Sure @halter73 . Then we need to use  `(string?)null or (Uri?)null` to select the right overload. Please correct If I am wrong

**Comment by halter73** [comment]
Yep.

**Comment by singh733** [comment]
Hi @halter73 

I have created [PR ](https://github.com/dotnet/aspnetcore/pull/42502/commits/6874170f9df8fc5c2b65eb07dbdf9c08a4336d3f)
Please let me know if there is any change.

**Comment by singh733** [comment]
@halter73  would you please provide any update on this.

**Comment by halter73** [comment]
Given the workarounds like `StatusCodeResult` or just setting the location even though it's unnecessary, this hasn't been my biggest priority. We are nearing code complete for .NET 7. I'm sorry. I still do plan to merge the PR, but it probably won't make it until .NET 8.

**Comment by halter73** [comment]
We should have been clearer about it, but the approved API is exactly
what's described in
https://github.com/dotnet/aspnetcore/issues/39454#issuecomment-1012457232.

**Comment by mitchdenny** [comment]
@captainsafia sorry I labelled this as mvc, but just noticed you had just labelled it mvc. It looks mostly MVC related to me but I guess we can use these types too in minimal.

**Comment by amis92** [comment]
It seems this has been closed by #42576
